### DISCUSSION TITLE
Closing toplevel value

### DISCUSSION THREAD

**Discussion by istvan-andrasi** [cue-lang/cue#3465] [Ideas]
Hi! I wanted to only allow fields matching a regex at the toplevel, and I tried this:
~~~
package tmp

#Top
#Top: [=~"^[A-Z_]+$"]: {...}

OK: {}
nok: {}
~~~

I expected it to fail, but it works:
~~~
> cue eval
#Top: {}
OK: {}
nok: {}
~~~

A assumed (and verified) it is because:
~~~
#Top     // is actually equivalent to
{ #Top } // which is embedding, thus allowing other fields
~~~

Although I found a workaround:
~~~
package tmp

[=~"^[A-Z_]+$"]: {...}
[!~"^[A-Z_]+$"]: _wrong_name: true & false

OK: {}
nok: {}
~~~
~~~
> cue eval
nok._wrong_name: conflicting values false and true:
    ./tmp.cue:7:22
    ./tmp.cue:7:35
    ./tmp.cue:7:42
~~~

I wanted to bring up if it would be sensible to treat closed toplevel struct references differently:
~~~
#Top     // would mean <toplevel>: #Def
{ #Top } // would mean <toplevel>: { #Def }
~~~

What do you think?

**Comment by mvdan** [comment]
This should be resolved with the current `explicitopen` experiment: https://github.com/cue-lang/cue/discussions/4032
